[MRG + 1] Added DisjointLabelKFold to perform K-Fold cv on sets with disjoint labels.

JeanKossaifi opened the pull request:
Added a SubjectIndependentKFold class to create subject independent folds.

JeanKossaifi committed:
Added subject independent KFold

amueller commented:
This is the same as LeavePLabelOut, right?
http://scikit-learn.org/dev/modules/generated/sklearn.cross_validation.LeavePLabelOut.html#sklearn.cross_validation.LeavePLabelOut
Maybe we should rename it, it is not super clear...

amueller commented:
SubjectIndependent is not a great name either, though, as it is very domain specific.

amueller commented:
Sorry I just now saw the discussion on the mailing list. If it is very similar to LeavePLabelOut, we indeed should either add a ``Stratified`` version or add a stratified parameter. We don't want domain-specific names.

JeanKossaifi commented:
Hi Andreas,

Thanks for the feedback.
My issue with adding a stratified parameter is that the name might be a bit
misleading (we don't want to leave P labels out, we just want disjoint
training and testing sets, while keeping approximately equilibrated folds).

Cheers,

Jean

2015-03-24 16:26 GMT+00:00 Andreas Mueller <notifications@github.com>:

> Sorry I just now saw the discussion on the mailing list. If it is very
> similar to LeavePLabelOut, we indeed should either add a Stratified
> version or add a stratified parameter. We don't want domain-specific names.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4444#issuecomment-85588789>
> .
>

amueller commented:
I think adding what you want seems like a good idea. I just don't have a good name. Maybe GroupIndependentKFold?

amueller commented:
sorry misclick.

ogrisel commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -1038,3 +1038,59 @@ def test_check_is_partition():
+        with no subject appearing in two different folds
+    """
+    # Fix the seed for reproducibility 
+    np.random.seed(0)
```
To avoid side effect between tests, please avoid seeding the global PRNG instance but instead use:

```python
rng = np.random.RandomState(0)

...
subjects = np.random.randint(0, n_subjects, n_samples)

```

ogrisel commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,71 @@ def __len__(self):
 
 
+def subject_independent_folds(subjects, n_folds=3):
+    """ Creates folds where a same subject is not in two different folds
```
cosmetics: no leading whitespace before "Creates".

ogrisel commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -1038,3 +1038,59 @@ def test_check_is_partition():
+    ideal_n_subjects_per_fold = n_samples // n_folds
+    
+    # Check that folds have approximately the same size
+    assert(len(folds)==len(subjects))
```
Please use `assert_equal` to get meaningful failure messages.

ogrisel commented:
I agree with @amueller with respect to the name. Baybe we could use `DisjointGroupKFold` instead? I find the "subject" naming too specific and the "independent" naming confusing / misleading.

JeanKossaifi committed:
Changed SubjectIndependentKFold to DisjointGroupKFold

cosmetic changes  test (fix seed correctly, use assert_equal for
meaningful error messages)

JeanKossaifi commented:
Thanks @ogrisel for the review!

Made the corrections and changed the name to DisjointGroupKFold.

jnothman commented:
I don't like the use of Group when elsewhere in CV Label actually means the
same thing. Group might be the better name, but we should be consistent.

On 25 March 2015 at 22:44, landscape-bot <notifications@github.com> wrote:

>  [image: Code Health] <https://landscape.io/diff/118267>
> Code quality remained the same when pulling *37ecdd7
> <https://github.com/JeanKossaifi/scikit-learn/commit/37ecdd7776f9293c19f233e23e0f2ddffab1d9e6>
> on JeanKossaifi:sid_kfold* into *6354e45
> <https://github.com/scikit-learn/scikit-learn/commit/6354e452db751a08fce3f8d50c0c945d04c1d7fd>
> on scikit-learn:master*.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4444#issuecomment-85993556>
> .
>

GaelVaroquaux commented:
> I don't like the use of Group when elsewhere in CV Label actually means the
> same thing. Group might be the better name, but we should be consistent.

+1

JeanKossaifi committed:
Changed name to DisjointLabelKFold

JeanKossaifi commented:
@jnothman, @GaelVaroquaux +1 for consistency
I changed the name to DisjointLabelKFold.

ragv commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,83 @@ def __len__(self):
+            The folds are built so that the same label doesn't appear in two different folds
+
+        n_folds: int, default is 3
+            number of folds
```
Could you add a nice `Examples` section like we have for `LeavePLabelOut` [here](https://github.com/scikit-learn/scikit-learn/blob/59efcb5/sklearn/cross_validation.py#L537)?

JeanKossaifi committed:
Added example of use

ragv commented:
Thanks for adding the example :)

( There is a whitespace related failure at https://travis-ci.org/scikit-learn/scikit-learn/jobs/55942000#L1336 )

JeanKossaifi committed:
FIX: whitespace related doctest failure

JeanKossaifi committed:
FIX: Python 2.6 requires the field numbers in print

ragv commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -1038,3 +1039,57 @@ def test_check_is_partition():
+def test_disjoint_label_folds():
+    """ Check that the function produces equilibrated folds
+        with no label appearing in two different folds
+    """
```
Could you also convert this to a comment following #4432 ? (sorry for the trouble)

ragv replied:
Thanks for the quick response!

JeanKossaifi replied:
Thanks, I didn't see that issue. Just changed the docstring into comment.

JeanKossaifi committed:
FIX: change docstring to comment in test function

JeanKossaifi committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into sid_kfold

JeanKossaifi committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into sid_kfold

BigCrunsh commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,109 @@ def __len__(self):
+
+
+class DisjointLabelKfold(_BaseKFold):
+    def __init__(self, y, n_folds=3):
```
I guess the description should be above of the function signature.

amueller replied:
Yeah it should be the class docstring, not the function docstring.

BigCrunsh commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,109 @@ def __len__(self):
+        """
+        self.n_folds = n_folds
+        self.n = len(y)
+        self.idxs = disjoint_label_folds(y=y, n_folds=n_folds)
```
Perhaps you should also call the super constructor here; it checks, e.g., if the arguments are valid:
```python
super(LabelSegmentedKFold, self).__init__(np.unique(labels).size, n_folds, shuffle, random_state)
```
It expects also ``shuffle`` and ``random_state``. Does it makes sense to extend the functionality in this direction?

BigCrunsh commented:
Hey @JeanKossaifi.
I think this iterator is very useful and I am looking forward to use it. @amueller just pointed me to that PR as I also started implementing it (#4709). Just two questions:
 - While I really like the idea that the folds are approximately balanced, do you think that the possibility of random disjoint splits might be useful, too?
 - Currently, the iterator returns empty test folds if ``n_folds > len(labels)``; this would be prevented if you call the super constructor.

Cheers

Christoph

JeanKossaifi commented:
@BigCrunsh thanks for the comments.

It is possible to add some degree of randomness in the current implementation by shuffling the labels that have the same weights (ie the appear the same number of time).

JeanKossaifi committed:
DOC: moved docstring from function to class

JeanKossaifi committed:
FIX: added call to parent class

JeanKossaifi committed:
FIX: error in calling the parent

JeanKossaifi committed:
DOC: fixed doctest

JeanKossaifi committed:
FIX: doctest

vene commented:
I'm all for consistency, but I think `Label` is a confusing name for both this and `Leave*LabelOut`. Many people call classes labels (even scikit-learn in e.g. multilabel).

amueller commented:
I think we should leave it for now. We could rename them in #4294?

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,112 @@ def __len__(self):
+    
+    Parameters
+    ----------
+    y: iterable of shape (n_samples, )
```
space before ``:``, no space after ``,`` I think (not sure on that though).

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,112 @@ def __len__(self):
+    Parameters
+    ----------
+    y: iterable of shape (n_samples, )
+        contains an id for each sample
```
Capitalize, full stop at the end.

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,112 @@ def __len__(self):
+
+class DisjointLabelKfold(_BaseKFold):
+    """Creates K approximately equilibrated folds
+        where the same label will not appear in two different folds
```
indentation

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,112 @@ def __len__(self):
+    unique_labels = np.unique(labels)
+    
+    # number of occurrence of each label (its "weight")
+    weight_per_label = sorted([(sum(labels == label), label) for label in unique_labels])
```
you should integer-code the labels by
```python
unique_labels, y = np.unique(y, return_inverse=True)
```
then you can just do
```python
weight_per_label = np.bincount(y)
```

amueller replied:
(your solution currently has quadratic complexity in the length of ``unique_labels``).

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,112 @@ def __len__(self):
+        ind_min = np.argmin(weight_per_fold)
+        w, label = weight_per_label.pop()
+        weight_per_fold[ind_min] += w
+        folds[labels == label] = ind_min
```
maybe instead of filling the folds as you go, just keep an array ``label_to_fold`` of length ``unique_labels`` that maps for each label which fold it will go to, and later do ``folds = label_to_fold[y]``

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,112 @@ def __len__(self):
+     [3 4]] [3 4] [1 2]
+    """
+
+    def __init__(self, y, n_folds=3):
```
just add a ``shuffle`` keyword ;)

vene commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,112 @@ def __len__(self):
+    return folds
+
+
+class DisjointLabelKfold(_BaseKFold):
```
I think the F should be capitalized for consistency

JeanKossaifi committed:
Cosmetic changes (minor refactoring)

JeanKossaifi committed:
Optimised code (use np.bincount)

JeanKossaifi commented:
@vene Changed to KFold for consistency.
@amueller made the changes you suggested + switched to np.bincount and used a label_to_fold array.
                   That was a great idea, thanks!

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,119 @@ def __len__(self):
+    weight_per_label = np.bincount(y)
+    # We want to distribute the heaviest weights first
+    ind = np.argsort(weight_per_label)[::-1]
+    weight_per_label = weight_per_label[ind]
```
why don't you just ``np.sort(weight_per_label)``? you don't need ind anywhere else, right?

amueller replied:
btw, why don't you call it sample_per_label instead of weight_per_label? I'm not sure what taking about weights adds.

JeanKossaifi replied:
@amueller I use ind at the end to keep the correspondence between the shuffled label and the original label.
You're right, I'll change the weight_per_label into sample_per_label, it is clearer! I just used 'weight' as I was reading about the knapsack problem at the time :)

amueller commented:
can you please add a test for the shuffle parameter?

JeanKossaifi committed:
Cosmetic: use samples instead of weight for clarity

JeanKossaifi committed:
Minor fix: removed shuffle parameter

amueller commented:
LGTM

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,119 @@ def __len__(self):
+        Contains an id for each sample.
+        The folds are built so that the same id doesn't appear in two different folds.
+    
+    n_folds: int, default is 3
```
space before :

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,119 @@ def __len__(self):
+        
+    Returns
+    -------
+    folds: numpy array of shape (n_samples, )
```
folds : numpy array, shape (n_samples,)

amueller commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,119 @@ def __len__(self):
+        Contains a label for each sample.
+        The folds are built so that the same label doesn't appear in two different folds.
+
+    n_folds : int, default is 3
```
int, default=3

amueller commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -1038,3 +1039,57 @@ def test_check_is_partition():
+    n_folds = 5
+    
+    # Construct the test data
+    tolerance = 0.05 * n_samples # 5 percent error allowed
```
where does this number come from?

JeanKossaifi replied:
I just used the same default as for the other classes (KFold, etc).

amueller replied:
You mean test_stratified_shuffle_split_even?

amueller replied:
That is quite a different test, though, right?

JeanKossaifi replied:
Oh sorry, I misread your question, I though you were referring to the default number of folds.
I just chose something seemingly reasonable (0.05 being usually used in statistics).

amueller replied:
It seems a bit magic to me, it is not clear why the test should pass. It depends on the distribution of labels, right? I guess with randint and many samples, they will be quite close.

JeanKossaifi replied:
Yes, this is true... I chose quite a large number of samples (1000) though and np.random.randint draws integers from a “discrete uniform” distribution in the “half-open” interval [low, high).
The seed is also fixed.

Should I leave that or do you suggest I simply manually write an array?

amueller replied:
I think it's fine for now.

amueller commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -1038,3 +1039,57 @@ def test_check_is_partition():
+        assert_greater_equal(tolerance, abs(sum(folds == i) - ideal_n_labels_per_fold))
+    
+    # Check that each subjects appears only in 1 fold
+    for label in np.unique(labels):
```
you could remove the for loop by constructing a coo_matrix instead, but I'm not sure it's worth it. this one might be cleaner.

JeanKossaifi replied:
I agree, to the reader it is probably easier to have the loopy version.

JeanKossaifi committed:
Cosmetic

ogrisel commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,119 @@ def __len__(self):
+    # number of occurrence of each label (its "weight")
+    samples_per_label = np.bincount(y)
+    # We want to distribute the most frequent labels first
+    ind = np.argsort(samples_per_label)[::-1]
```
I wonder if we should not use a stable stort algorithm like `np.argsort(samples_per_label, kind="mergesort")` here.

@larsmans any idea if a non-stable sort could cause reproducibility issues in this case?

larsmans replied:
If the NumPy implementation of quicksort changes, you can get a different ordering for tied labels. From a quick glance at the code, I don't see how that would not affect the output.

Use mergesort.

larsmans commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,119 @@ def __len__(self):
+    ----------
+    y: numpy array, shape (n_samples,)
+        Contains an id for each sample.
+        The folds are built so that the same id doesn't appear in two different folds.
```
id -> index

And maybe document the dtype; it should be `np.intp`.

JeanKossaifi replied:
@larsmans @ogrisel changed to mergesort, thanks.
@larsmans currently the implementation allows the user to have an array of strings for y (for instance the name or id of the subjects in an experiment) as I didn't see any reason to not allow that. Do you want to enforce a np.intp dtype?

jnothman replied:
I think string should be welcome.

JeanKossaifi committed:
Use mergesort instead of quicksort for reproducibility.

jnothman commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,119 @@ def __len__(self):
         return self.n_folds
 
 
+def disjoint_label_folds(y, n_folds=3):
```
I'm sure I missed discussion somewhere, but why did this get called `y`?

JeanKossaifi replied:
We could call it labels, y is simply for consistency with other cross-validation methods (.e.g. StratifiedKFold).

jnothman replied:
But `y` means something different there; there it is the target of prediction over which samples are stratified. This variable is much more like `label in `LeaveOneLabelOut`

ogrisel replied:
+1, you cannot predict something you have never seen on the training set. `y` in scikit-learn is always the target variable for supervised prediction. I agree, it's a bad name but it's too late to change.

JeanKossaifi replied:
@jnothman @ogrisel ok, changed y to labels :)

JeanKossaifi committed:
Changed variable name 'y' to 'label'.

jnothman commented:
I'm trying to think of degenerate cases. We should raise an error if `n_labels < n_folds`, or else we'll produce an empty test set. Perhaps we should also say in the documentation that this will work best if `n_labels >> n_folds` and the samples are relatively evenly spread among the folds.

JeanKossaifi committed:
Added test for degenerate case where n_folds > n_labels.

JeanKossaifi committed:
Documented the requirement n_labels > n_folds.

JeanKossaifi commented:
@jnothman good point, I added a test.

ogrisel commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,126 @@ def __len__(self):
+
+
+class DisjointLabelKFold(_BaseKFold):
+    """Creates K approximately equilibrated folds.
```
I would not put emphasis on balancing but rather on mutual label exclusion:

```python
class DisjointLabelKFold(_BaseKFold):
    """K-fold iterator variant with non-overlapping labels.

    The same label will not appear in two different folds  (the number of
    labels has to be at least equal to the number of folds).

    The folds are approximately balanced in the sense so that the number of
    distinct labels is approximately the same in each fold.
    """
```

amueller replied:
On 07/01/2015 12:46 PM, Olivier Grisel wrote:
>
> I would not put emphasis on balancing but rather on mutual label 
> exclusion:
>
We also have that in #4583 though, right?

ogrisel commented:
Please mention this new class in the "see also" section of the `KFold`, `StratifiedKFold`, `LeaveOneLabelOut` and `LeavePLabelOut` class docstrings.

Please also introduce this tool in the user guide in a new section in `sklearn/doc/modules/cross_validation.rst`, probably after the section on `StratifiedKFold`.

You also need to add a new entry for this class at the appropriate location in `sklearn/doc/modules/classes.rst` (reference documentation).

ogrisel commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,126 @@ def __len__(self):
+    samples_per_fold = np.zeros(n_folds)
+
+    # Mapping from label index to fold index
+    label_to_fold = np.zeros(len(unique_labels))
```
I think this should be `np.zeros(len(unique_labels), dtype=np.uintp)` instead.

ogrisel commented on sklearn/cross_validation.py:
```diff
@@ -332,6 +332,126 @@ def __len__(self):
+    samples_per_label = samples_per_label[ind]
+
+    # Total weight of each fold
+    samples_per_fold = np.zeros(n_folds)
```
I think this should be `np.zeros(len(unique_labels), dtype=np.uint64)` instead.

ogrisel commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -1038,3 +1039,60 @@ def test_check_is_partition():
+    
+    # Construct the test data
+    tolerance = 0.05 * n_samples # 5 percent error allowed
+    labels = np.random.randint(0, n_labels, n_samples)
```
Please use the isolated, seeded `rng` instance instead of the `np.random` singleton.

ogrisel commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -1038,3 +1039,60 @@ def test_check_is_partition():
+    
+    # Check that each subjects appears only in 1 fold
+    for label in np.unique(labels):
+        assert_equal(len(np.unique(folds[labels == label])), 1)
```
Please enhance this test by also checking that using the `DisjointLabelKFold` iterator you never have samples with the same label on both sides of the split, e.g.:

```python

labels = np.asarray(labels, dtype=object)  # to allow fancy indexing on labels

for train, test in cval.DisjointLabelKFold(labels, n_folds=n_folds):
    assert_equal(len(np.intersect1d(labels[train], labels[test])), 0)
```

JeanKossaifi committed:
DOC: improved description + added see also sections.

JeanKossaifi committed:
Fixed dtype of temporary arrays.

JeanKossaifi committed:
Improved test: check that one label is not in both test and training.

JeanKossaifi committed:
Added documentation for DisjoinLabelKFold.

JeanKossaifi commented:
Thanks @ogrisel, I added the documentation and addressed the other issues.

glouppe commented:
I know there was already some discussion and changes regarding the naming of this iterator, but what about simply `LabelKFold`? 

(We just added `LabelShuffleSplit` as a label-variant of `ShuffleSplit`. Hence, `LabelKFold` would be the label-variant of `KFold`)

GaelVaroquaux commented:
> I know there was already some discussion and changes regarding the naming of
> this iterator, but what about simply LabelKFold?

I would be happy with that.

jnothman commented:
I know we're stuck with "Label" for now, but "GroupedKFold" is much clearer!

On 30 August 2015 at 23:32, Gael Varoquaux <notifications@github.com> wrote:

> > I know there was already some discussion and changes regarding the
> naming of
> > this iterator, but what about simply LabelKFold?
>
> I would be happy with that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4444#issuecomment-136138490>
> .
>

agramfort commented:
I like LabelKFold :)

glouppe commented:
I am taking care of rebasing and renaming.
